applications/openbrush-follow-up.md

```diff
@@ -0,0 +1,287 @@
+| -----  | -----------                                                              | -------------                                                                                                                             |
+| 0a.    | License                                                                  | MIT                                                                                                                                       |
+| 0b.    | Documentation                                                            | We will provide inline documentation, example of usage of extensions. |
+| 0c.    | Testing Guide                                                            | We will add tests for extensions and for a new changes from ink! side.                                                           |
+| 1.     | Contribute to ink! with fixing of events                                 | We will help to fix the [issue](https://github.com/paritytech/ink/issues/809) with events. |
+| 2.     | Add support of default implementation in trait definition on ink! level  | We will help with the support of default implementation inside of trait definition. It will require discussions with the ink! team to define the best way how to implement that without conflicts with their future changes.  |
+| 3.     | Refactor of implementation according changes in ink!                     | After changes in ink! we will refactor the code of library.  |
+
+#### Milestone 6 - Support of upgradable contracts
+
+* **Estimated Duration:** 7 weeks(time depends on conversations in ink! and `contract-pallet` repositories)
+* **FTE:**  2.5
+* **Costs:** 35,000 USD
+
+| Number | Deliverable                      | Specification                                                                                                                             |
+| -----  | -----------                      | -------------                                                                                                                             |
+| 0a.    | License                          | MIT                                                                                                                                       |
+| 0b.    | Documentation                    | We will provide inline documentation, example of upgradable contracts. |
+| 0c.    | Testing Guide                    | We will add tests to cover upgradability of contracts.                                                           |
+| 1.     | Implement delegated call         | We will find and provide the idea of how [delegeted call](https://docs.soliditylang.org/en/v0.4.21/introduction-to-smart-contracts.html#delegatecall-callcode-and-libraries) can be implemented in `contract-pallet`. Help with it's implementation.            |
```

Comment:
Why is the `delegatecall` needed for upgrade ability? AFAIK it is only needed for library contracts which are not relevant here. Correct me if I am wrong. I was under the impression that https://github.com/paritytech/substrate/pull/8909 implemented everything needed for upgradeable contracts.

Reply:
The idea of an upgradeable contract in Solidity, that you will execute the code of another contract in the context of the current contract(you will affect the storage of the current contract by the code of callee contract). 

This change doesn't allow that. It allows only to call another without the ability to return back by the `TAIL_CALL` flag.

Reply:
I know what delegatecall is. I just didn't know it was used for upgradeability. But it makes sense I guess. But you need to make sure that the new version does not change the storage layout, right?

Reply:
Yes, the developer of the upgradable contract must worry about layout. We will create a list of rules that the developer must follow in order not to break the layout. [OpenZeppelin](https://docs.openzeppelin.com/upgrades-plugins/1.x/writing-upgradeable) did the same for solidity. Our library will provide standards in upgradable form too.